### DISCUSSION TITLE
Неудобный переход по ссылкам на документацию, в которых есть конкретный header (h1, h2, ..) на странице

### DISCUSSION THREAD

**Discussion by simplepersonru** [totumonline/totum-ru-issues-and-discussions#388] [ОШБК!] [answered]
Возьмем для примера https://ru.docs.totum.online/tables#table-settings-panel. В данном случае header - **table-settings-panel**
Ожидаю что всегда при переходе по ссылке будем попадать на нужное место в документе
Видео:

https://github.com/totumonline/totum-ru-issues-and-discussions/assets/71624171/fa7e4cee-6ba6-44c6-91de-70a5b9824005

И получается, что тратится много дополнительного времени чтобы перейти на нужное место в документации

Я уже сталкивался с этим. В некоторых проектах есть эта проблема, в некоторых нет. Кто-то из коллег предполагал что это связано с кешированием в браузере. Если есть возможность как-то это порешать для общего случая обычного пользователя, было бы славно

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Мы делали около 5 подходов к этой задаче (ни один не помог). Если вы дадите ссылку на метод, который это фиксит - всем станет лучше.

UPD: вроде бы починили, но это не точно...

**Reply by simplepersonru** [reply]
я очень очень мало знаком с web-разработкой, но вот попытка решения в лоб, если предположить что дело в асинхронности загрузки ресурсов и в частности высоких картинок, которые сильно могут сбить прицел скролла. Проверить в боевом проекте возможности и времени нет, но из консоли разработчика работает. Надеюсь поможет. 

Соответственно если у нас React (а у вас я видел React на фронте), то функция magic идет в useEffect (или ComponentDidMount для классовых компонентов). Этот useEffect должен быть в компонентах, которые являются маршрутом (Route) в статьи документации

```
function getHeading() {
    let currentURL = window.location.href;
    let hashIndex = currentURL.indexOf('#');
    if (hashIndex !== -1) {
        return currentURL.substring(hashIndex + 1);
    }
    else {
        return null;
    }
}

function magic() {
    const heading = getHeading()
    if (heading == null) //Если в ссылке нет заголовка, ничего не делаем
        return

    //Подписка на полную загрузку всего документа
    document.addEventListener('load', function () {
        // Код, который будет выполнен после полной загрузки всех ресурсов документа

        // Раскоментить если не работает - дополнительное небольшое хардкод ожидание после загрузки
        //await new Promise(r => setTimeout(r, 500)); 

        var hElement = document.querySelector('#' + heading);
        window.scrollTo({
            top: hElement.offsetTop,
            behavior: 'smooth' // Добавляет плавную анимацию прокрутки, если нужно
        });
    });
}
```

**Reply by simplepersonru** [reply]
@ttm-alexeypolunin Если будет возможность проверить, дайте пожалуйста обратную связь

**Reply by ttm-alexeypolunin** [reply]
Поправили по вашей наводке неделю назад. Но у меня самого не было еще возможности удостовериться, что точно работает.

**Reply by simplepersonru** [reply]
попробовал, не работает. Значит, возможно, я не угадал)

**Reply by ttm-alexeypolunin** [reply]
Мы что-то такое уже делали, немного по другому но логика была такая же

**Reply by simplepersonru** [reply]
возможно моя реализация на ванильном JS не учитывает то, что React работает с виртуальным DOM. Могу сделать наивное предположение, что ванильный сигнал load не окончательный в загрузке страницы, при использовании фреймворка React
